### ISSUE TITLE
Enabling ssa-claims feature flag in crossplane-1.15 on a cluster with existing claims is not upgrade-safe

### ISSUE THREAD

**Issue by ravilr** [crossplane/crossplane#5498]
### What happened?

On enabling [ssa-claims flag](https://docs.crossplane.io/latest/concepts/server-side-apply/#use-server-side-apply-to-sync-claims-with-composite-resources), all the existing XRs(which were successfully reconciled and composed before) are losing Composite's existing `spec.resourceRefs[]` and composite reconciler creating new composed resources from the composition rendering and overwrites the Composite's `spec.resourceRefs` with the new composed resources references. this is resulting in duplicate composed resources, one set (new ones) which are tracked by the XR resourceRefs, and the old composed resources which the XR no more knows about. 

This seems to be due to the way managedFields ownership are being updated on ssa-claim feature enablement.  managedField ownership of `spec.resourceRefs` field of an XR is being updated to `manager: apiextensions.crossplane.io/claim`, but the first SSA update/patch doesn't include `spec.resourceRefs` in the SSA patch update request, resulting in those fields removed from the API ?

### How can we reproduce it?

- Create a claim before enabling ssa-claim: https://github.com/crossplane/crossplane/blob/40f0a2909646524879e4d1d4597b79eb138e4ca8/test/e2e/manifests/lifecycle/upgrade/claim.yaml

```
$ kubectl get composite,managed -l crossplane.io/claim-name=lifecycle-upgrade
NAME                                                   SYNCED   READY   COMPOSITION                     AGE
xnopresource.nop.example.org/lifecycle-upgrade-tn26x   True     True    xnopresources.nop.example.org   2m41s

NAME                                                          READY   SYNCED   AGE
nopresource.nop.crossplane.io/lifecycle-upgrade-tn26x-42h8q   True    True     2m52s
```

composite XR with `--show-managed-fields` with CSA:

```
apiVersion: nop.example.org/v1alpha1
kind: XNopResource
metadata:
  creationTimestamp: "2024-03-21T22:13:23Z"
  finalizers:
  - composite.apiextensions.crossplane.io
  generateName: lifecycle-upgrade-
  generation: 4
  labels:
    crossplane.io/claim-name: lifecycle-upgrade
    crossplane.io/claim-namespace: default
    crossplane.io/composite: lifecycle-upgrade-tn26x
  managedFields:
  - apiVersion: nop.example.org/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"composite.apiextensions.crossplane.io": {}
        f:generateName: {}
        f:labels:
          .: {}
          f:crossplane.io/claim-name: {}
          f:crossplane.io/claim-namespace: {}
          f:crossplane.io/composite: {}
      f:spec:
        .: {}
        f:claimRef:
          .: {}
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
        f:compositionRef:
          .: {}
          f:name: {}
        f:compositionRevisionRef:
          .: {}
          f:name: {}
        f:compositionUpdatePolicy: {}
        f:coolField: {}
        f:resourceRefs: {}
    manager: crossplane
    operation: Update
    time: "2024-03-21T22:13:23Z"
  - apiVersion: nop.example.org/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:conditions:
          .: {}
          k:{"type":"Ready"}:
            .: {}
            f:lastTransitionTime: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"Synced"}:
            .: {}
            f:lastTransitionTime: {}
            f:reason: {}
            f:status: {}
            f:type: {}
    manager: crossplane
    operation: Update
    subresource: status
    time: "2024-03-21T22:13:48Z"
  name: lifecycle-upgrade-tn26x
  resourceVersion: "28726594"
  uid: d1d24563-19cc-4919-b82b-a3abf291f290
spec:
  claimRef:
    apiVersion: nop.example.org/v1alpha1
    kind: NopResource
    name: lifecycle-upgrade
    namespace: default
  compositionRef:
    name: xnopresources.nop.example.org
  compositionRevisionRef:
    name: xnopresources.nop.example.org-04c418e
  compositionUpdatePolicy: Automatic
  coolField: I'm cool!
  resourceRefs:
  - apiVersion: nop.crossplane.io/v1alpha1
    kind: NopResource
    name: lifecycle-upgrade-tn26x-42h8q
status:
  conditions:
  - lastTransitionTime: "2024-03-21T22:13:23Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
  - lastTransitionTime: "2024-03-21T22:13:48Z"
    reason: Available
    status: "True"
    type: Ready
```

- Enable `--enable-ssa-claims` flag in crossplane deployment 

- Notice that the XR's resourceRefs are reset and new composed resources are created:

```
$ kubectl get composite,managed -l crossplane.io/claim-name=lifecycle-upgrade
NAME                                                   SYNCED   READY   COMPOSITION                     AGE
xnopresource.nop.example.org/lifecycle-upgrade-tn26x   True     True    xnopresources.nop.example.org   10m

NAME                                                          READY   SYNCED   AGE
nopresource.nop.crossplane.io/lifecycle-upgrade-tn26x-42h8q   True    True     10m
nopresource.nop.crossplane.io/lifecycle-upgrade-tn26x-lf4xh   True    True     3m17s # <=====  New MR composed, which isn't supposed to be.
```

composite XR with `--show-managed-fields` after ssa reconcile:
```
apiVersion: nop.example.org/v1alpha1
kind: XNopResource
metadata:
  creationTimestamp: "2024-03-21T22:13:23Z"
  finalizers:
  - composite.apiextensions.crossplane.io
  generation: 7
  labels:
    crossplane.io/claim-name: lifecycle-upgrade
    crossplane.io/claim-namespace: default
    crossplane.io/composite: lifecycle-upgrade-tn26x
  managedFields:
  - apiVersion: nop.example.org/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:labels:
          f:crossplane.io/claim-name: {}
          f:crossplane.io/claim-namespace: {}
      f:spec:
        f:claimRef:
          f:apiVersion: {}
          f:kind: {}
          f:name: {}
          f:namespace: {}
        f:compositionRef:
          f:name: {}
        f:compositionUpdatePolicy: {}
        f:coolField: {}
    manager: apiextensions.crossplane.io/claim
    operation: Apply
    time: "2024-03-21T22:20:25Z"
  - apiVersion: nop.example.org/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .: {}
          v:"composite.apiextensions.crossplane.io": {}
        f:labels:
          f:crossplane.io/composite: {}
      f:spec:
        f:compositionRevisionRef:
          .: {}
          f:name: {}
        f:resourceRefs: {}
    manager: crossplane
    operation: Update
    time: "2024-03-21T22:20:25Z"
  - apiVersion: nop.example.org/v1alpha1
    fieldsType: FieldsV1
    fieldsV1:
      f:status:
        .: {}
        f:conditions:
          .: {}
          k:{"type":"Ready"}:
            .: {}
            f:lastTransitionTime: {}
            f:reason: {}
            f:status: {}
            f:type: {}
          k:{"type":"Synced"}:
            .: {}
            f:lastTransitionTime: {}
            f:reason: {}
            f:status: {}
            f:type: {}
    manager: crossplane
    operation: Update
    subresource: status
    time: "2024-03-21T22:20:27Z"
  name: lifecycle-upgrade-tn26x
  resourceVersion: "28730378"
  uid: d1d24563-19cc-4919-b82b-a3abf291f290
spec:
  claimRef:
    apiVersion: nop.example.org/v1alpha1
    kind: NopResource
    name: lifecycle-upgrade
    namespace: default
  compositionRef:
    name: xnopresources.nop.example.org
  compositionRevisionRef:
    name: xnopresources.nop.example.org-04c418e
  compositionUpdatePolicy: Automatic
  coolField: I'm cool!
  resourceRefs:
  - apiVersion: nop.crossplane.io/v1alpha1
    kind: NopResource
    name: lifecycle-upgrade-tn26x-lf4xh
status:
  conditions:
  - lastTransitionTime: "2024-03-21T22:13:23Z"
    reason: ReconcileSuccess
    status: "True"
    type: Synced
  - lastTransitionTime: "2024-03-21T22:20:27Z"
    reason: Available
    status: "True"
    type: Ready
```

### What environment did it happen in?
Crossplane version: v1.15.1

* Kubernetes version: 
```
Client Version: version.Info{Major:"1", Minor:"27", GitVersion:"v1.27.3", GitCommit:"25b4e43193bcda6c7328a6d147b1fb73a33f1598", GitTreeState:"clean", BuildDate:"2023-06-14T09:47:38Z", GoVersion:"go1.20.5", Compiler:"gc", Platform:"darwin/arm64"}
Kustomize Version: v5.0.1
Server Version: version.Info{Major:"1", Minor:"27+", GitVersion:"v1.27.10-eks-508b6b3", GitCommit:"e99f7c75641f738090d483d988dc4a70001e01cf", GitTreeState:"clean", BuildDate:"2024-01-29T20:59:05Z", GoVersion:"go1.20.13", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by ravilr** [comment]
> This seems to be due to the way managedFields ownership are being updated on ssa-claim feature enablement. managedField ownership of `spec.resourceRefs` field of an XR is being updated to `manager: apiextensions.crossplane.io/claim`, but the first SSA update/patch doesn't include `spec.resourceRefs` in the SSA patch update request, resulting in those fields removed from the API ?

@negz @phisco @turkenh  please review below analysis and suggest how do we want to proceed:

before --enable-ssa-claims feature flag is enabled, 
1.) In case of Composite/XR's reconciled by the native PTComposer, the `spec.resourceRef` field is owned by the singular csa based manager `crossplane` in the metadata.managedFields entries.  These run into the issue reported here on ssa-claims enabling.
2.) In case of Composite/XR's reconciled by the Function based Composer,  the `spec.resourceRef` field is owned by the ssa based manager `apiextensions.crossplane.io/composite`.  these don't run into the issue reported here.

on enabling ssa-claims flag, in case of (1.) above,
the metadata.managedFields ownership transfer of an XR/Composite from csa based manager `crossplane` to ssa based manager `apiextensions.crossplane.io/claim` by the claim reconciler at https://github.com/crossplane/crossplane/blob/v1.15.1/internal/controller/apiextensions/claim/syncer_ssa.go#L102 results in `spec.resourceRefs` field to be owned by the the `apiextensions.crossplane.io/claim` manager entry in the metadata.managedFields.  

after this ownership transfer, the SSA patch request applied on XR by the claim reconciler at https://github.com/crossplane/crossplane/blob/v1.15.1/internal/controller/apiextensions/claim/syncer_ssa.go#L252 doesn't include the `spec.resourceRefs`  in the xrPatch object spec, resulting in that getting removed from the live object.  At this point, the Composite reconciler notices empty resourceReferences and goes on to compose new set of resources and adds them to the resourceRefs.

Since the claim reconciler have no opinion on XR's `spec.resourceRefs`,  the initial ownership transfer isn't expected to include `spec.resourceRefs` in the managedFields entries owned by the new ssa claim manager..

I tested this by stripping the `spec.resourceRefs` field from the metadata.managedFields entries patch returned by `csaupgrade.UpgradeManagedFieldsPatch(..)`,  before the upgrade patch request in https://github.com/crossplane/crossplane/blob/v1.15.1/internal/controller/apiextensions/claim/syncer_ssa.go#L102. with that, the `spec.resourceRefs` stays intact in the live object on subsequent SSA apply by the claim reconciler.  the composite reconciler eventually takes over the field ownership of `spec.resourceRefs` in metadata.managedFields, whenever the sync results in updates/changes to resourceRefs.

```
diff --git a/internal/controller/apiextensions/claim/syncer_ssa.go b/internal/controller/apiextensions/claim/syncer_ssa.go
index 18d4f405..1ec2da6c 100644
--- a/internal/controller/apiextensions/claim/syncer_ssa.go
+++ b/internal/controller/apiextensions/claim/syncer_ssa.go
@@ -17,13 +17,17 @@ limitations under the License.
 package claim
 
 import (
+       "bytes"
        "context"
+       "encoding/json"
        "fmt"
 
+       metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/types"
        "k8s.io/apimachinery/pkg/util/sets"
        "k8s.io/client-go/util/csaupgrade"
        "sigs.k8s.io/controller-runtime/pkg/client"
+       "sigs.k8s.io/structured-merge-diff/v4/fieldpath"
 
        xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
        "github.com/crossplane/crossplane-runtime/pkg/errors"
@@ -99,7 +103,80 @@ func (u *PatchingManagedFieldsUpgrader) Upgrade(ctx context.Context, obj client.
                // No patch means there's nothing to upgrade.
                return nil
        }
-       return errors.Wrap(resource.IgnoreNotFound(u.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), errPatchFieldManagers)
+       filteredPatch, err := stripManagedFields(p)
+       if err != nil {
+               return errors.Wrap(err, errCreatePatch)
+       }
+       return errors.Wrap(resource.IgnoreNotFound(u.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, filteredPatch))), errPatchFieldManagers)
+}
+
+func stripManagedFields(patch []byte) ([]byte, error) {
...
```

**Comment by ravilr** [comment]
Opened a draft PR https://github.com/crossplane/crossplane/pull/5550 with the change proposed in https://github.com/crossplane/crossplane/issues/5498#issuecomment-2035944002,  for feedback.

**Comment by negz** [comment]
Awesome analysis, thanks @ravilr! Let me dig in a bit and get back to you.

**Comment by negz** [comment]
For context, the reason we try to upgrade the field managers in the first place is to avoid https://github.com/kubernetes/kubernetes/issues/99003.

If we didn't do the upgrade the new `apiextensions.crossplane.io/claim` field manager would share field ownership with the old `crossplane` field manager. This means that `apiextensions.crossplane.io/claim` omitting a field from its fully-specified intent wouldn't result in the field being deleted, because the API server believes `crossplane` still wants the field to exist.

For us, the challenge is that the fields owned by the `crossplane` field manager at upgrade-time actually represent the state managed by two controller - the claim controller and the XR controller. We only want to upgrade the bits owned by the claim controller.

I'd prefer to avoid not having to "do surgery" on the managed fields, but it seems like it might be unavoidable.

I wondered whether we could simply not upgrade the managed fields at all and clear them per [this documentation](https://kubernetes.io/docs/reference/using-api/server-side-apply/#clearing-managedfields). What that seems to do is create a new field manager called `before-first-apply` that owns field that existed before the first time (with operation type `Update`). So the claim controller ends up sharing ownership with `before-first-apply`.

**Comment by negz** [comment]
I played around a little further and opened a PR with another potential solution: https://github.com/crossplane/crossplane/pull/5555. @ravilr let me know what you think.

**Comment by ravilr** [comment]
I agree the approach in #5555 seems simpler and prefer it, if that gets us the intended behavior.

**Comment by ravilr** [comment]
Any chance could we include the fix for this in [release-1.15.2](https://github.com/crossplane/release/issues/7) ?

**Comment by jbw976** [comment]
Given that the two PRs/ideas to address this issue are still in draft, I'm going to go ahead and get started with the v1.14/v1.15 patch releases now @ravilr. We can always do another patch release for v1.15 with the fix for this issue once we land on the right solution and feel good about it 😉

**Comment by ravilr** [comment]
@negz  i've tested the patch from your #5555 PR, and found it to be working as desired in resolving the issue reported here in the claim reconciler. I've closed my PR in favor of yours, as it is simpler and don't require additional dependencies.

While testing this with the goal of end-end SSA (claim -> XR -> MRs),  have run into similar issue in Composite reconciler migration from native PT composition to Functions based composition, where the composed resources ends up having shared ownership of fields between `crossplane` and `apiextensions.crossplane.io/composed/XR_hash`, which results in any field removals in XR -> MRs to not sync/propagate as expected, after the functions migration..

**Comment by negz** [comment]
Thanks @ravilr! I'm just getting back from a week off (followed by a brief cold) so I've got a bit of a backlog. I'll aim to follow up on this early next week.

**Comment by ravilr** [comment]
> While testing this with the goal of end-end SSA (claim -> XR -> MRs), have run into similar issue in Composite reconciler migration from native PT composition to Functions based composition, where the MRs/composed resources ends up having shared ownership of fields between `crossplane` and `apiextensions.crossplane.io/composed/XR_hash` field managers, which results in any field removals in XR -> MRs to not sync/propagate as expected, after the functions migration..

@negz while #5555 addresses the field managers upgrade in `XR` by claim reconciler,  we found out that for Seamless migration to end-end SSA, the composite function reconciler also needs similar clearing of field managers in `composed` resources while migrating existing resources from Native P&T based compositions to function pipeline based compositions. We're running an additional patch internally on clusters with existing resources, to clear managed fields from all observed `composed` resources in FunctionComposer Compose():

```
diff --git a/internal/controller/apiextensions/composite/composition_functions.go b/internal/controller/apiextensions/composite/composition_functions.go
index a27372a5..4de24093 100644
--- a/internal/controller/apiextensions/composite/composition_functions.go
+++ b/internal/controller/apiextensions/composite/composition_functions.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"reflect"
 	"sort"
+	"strings"
 
 	"google.golang.org/protobuf/encoding/protojson"
 	"google.golang.org/protobuf/types/known/structpb"
@@ -454,6 +455,19 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 		return CompositionResult{}, errors.Wrap(err, errApplyXRRefs)
 	}
 
+	// TODO: Remove this call to clearManagedFields once no supported version of
+	// Crossplane uses native P&T. We only need to clear/upgrade field managers if the
+	// native ComposerPT might have applied the composed resources using the
+	// default client-side apply field manager "crossplane" before,
+	// but now migrated to use Composition functions, which uses server-side apply instead.
+	// Without this managedFields upgrade, the composed resources ends up having shared ownership
+	// of fields and field removals won't sync properly.
+	for _, cd := range observed {
+		if err := c.clearManagedFields(ctx, cd.Resource); err != nil {
+			return CompositionResult{}, errors.Wrap(err, "cannot clear composed resource's managed fields from client-side to server-side apply")
+		}
+	}
+
 	// Produce our array of resources to return to the Reconciler. The
 	// Reconciler uses this array to determine whether the XR is ready.
 	resources := make([]ComposedResource, 0, len(desired))
@@ -524,6 +538,75 @@ func (c *FunctionComposer) Compose(ctx context.Context, xr *composite.Unstructur
 	return CompositionResult{ConnectionDetails: d.GetComposite().GetConnectionDetails(), Composed: resources, Events: events}, nil
 }
 
+// Clear/Upgrade the supplied composed object's field managers from client-side to server-side
+// apply.
+//
+// This is a multi-step process.
+//
+// Step 1: All fields are owned by manager 'crossplane' operation 'Update'. This
+// represents all fields set by the XR controller up to this point.
+//
+// Step 2: Upgrade is called for the first time. We delete all field managers.
+//
+// Step 3: The XR controller server-side applies its fully specified intent
+// as field manager with prefix 'apiextensions.crossplane.io/composed/'. This becomes the
+// manager of all the fields that are part of the claim controller's fully
+// specified intent. All existing fields the XR ontroller didn't specify
+// become owned by a special manager - 'before-first-apply', operation 'Update'.
+//
+// Step 4: Upgrade is called for the second time. It deletes the
+// 'before-first-apply' field manager entry. Only the XR composed field manager
+// remains.
+func (c *FunctionComposer) clearManagedFields(ctx context.Context, obj client.Object) error {
+	// The composed resource doesn't exist, nothing to upgrade.
+	if !meta.WasCreated(obj) {
+		return nil
+	}
+
+	foundSSA := false
+	foundBFA := false
+	idxBFA := -1
+
+	for i, e := range obj.GetManagedFields() {
+		if strings.HasPrefix(e.Manager, FieldOwnerComposedPrefix) {
+			foundSSA = true
+		}
+		if e.Manager == "before-first-apply" {
+			foundBFA = true
+			idxBFA = i
+		}
+	}
+
+	switch {
+	// If our SSA field manager exists and the before-first-apply field manager
+	// doesn't, we've already done the upgrade. Don't do it again.
+	case foundSSA && !foundBFA:
+		return nil
+
+	// We found our SSA field manager but also before-first-apply. It should now
+	// be safe to delete before-first-apply.
+	case foundSSA && foundBFA:
+		p := []byte(fmt.Sprintf(`[
+			{"op": "remove", "path": "/metadata/managedFields/%d"},
+			{"op": "replace", "path": "/metadata/resourceVersion", "value": "%s"}
+		]`, idxBFA, obj.GetResourceVersion()))
+		return errors.Wrap(resource.IgnoreNotFound(c.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), "cannot remove before-first-apply from field managers")
+
+	// We didn't find our SSA field manager or the before-first-apply field
+	// manager. This means we haven't started the upgrade. The first thing we
+	// want to do is clear all managed fields. After we do this we'll let our
+	// SSA field manager apply the fields it cares about. The result will be
+	// that our SSA field manager shares ownership with a new manager named
+	// 'before-first-apply'.
+	default:
+		p := []byte(fmt.Sprintf(`[
+			{"op": "replace", "path": "/metadata/managedFields", "value": [{}]},
+			{"op": "replace", "path": "/metadata/resourceVersion", "value": "%s"}
+		]`, obj.GetResourceVersion()))
+		return errors.Wrap(resource.IgnoreNotFound(c.client.Patch(ctx, obj, client.RawPatch(types.JSONPatchType, p))), "cannot clear field managers")
+	}
+}
+
 // ComposedFieldOwnerName generates a unique field owner name
 // for a given Crossplane composite resource (XR). This uniqueness is crucial to
 // prevent multiple XRs, which compose the same resource, from continuously
```
